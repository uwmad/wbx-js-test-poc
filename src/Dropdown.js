import React from 'react';
import PropTypes from 'prop-types';
import WebixHelper from "./webix/WebixHelper";

const{ webix } = window;

const Dropdown = (props) => {

    const { options, onSelect, placeholder, disabled, width, id, value } = props;

    const handleChange = (newVal) => {
        let selected = '';
        options.forEach((option) => {
            if (option.id === newVal) {
                selected = option.value;
            }
        });

        onSelect(selected, id);
    };

    const config = {
        view:"combo",
        id,
        scroll:false,
        disabled,
        placeholder,
        width,
        height: 40,
        options,
        value,
        icon:"wxi-angle-down",
        on: {
            onChange(newVal) {
                handleChange(newVal);
            },
        },
    };

    const init = (helper) => webix.ui({
            ...config,
            container: helper.ref.current,
        });

    return(<WebixHelper webixInitFn={init} config={config} />);
};

Dropdown.propTypes = {
    options:        PropTypes.array,
    width:          PropTypes.number.isRequired,
    onSelect:       PropTypes.func.isRequired,
    id:             PropTypes.string.isRequired,
    placeholder:    PropTypes.string,
    disabled:       PropTypes.bool,
    value:          PropTypes.any,
};

Dropdown.defaultProps = {
    disabled:       false,
    placeholder:    "Select",
    value:          undefined,
    options:        [],
};

export default Dropdown;
