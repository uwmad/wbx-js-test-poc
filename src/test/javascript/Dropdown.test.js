import React from 'react';
import { mount } from 'enzyme';
import Dropdown from "../../Dropdown";
import waitForExpect from 'wait-for-expect';
import webixFromSelector from '../test_utils/webixFromSelector';

describe('Dropdown', () => {


    describe('Rendering', () => {

        const App = document.getElementById('App');
        const webixUISpy = jest.spyOn(webix, 'ui');
        let wrapper;

        const options = ['item1','item2','item3','item4'];
        const width = 100;
        const onSelect = jest.fn();
        const placeholder = 'placeholder text';
        const disabled = false;
        const value = 'item3';

        beforeEach(() => {
            wrapper = mount (
                    <Dropdown 
                        options = { options }
                        width = { width }
                        onSelect = { onSelect }
                        placeholder = { placeholder }
                        disabled = { disabled }
                        value = { value }
                    />,
                    { 
                        attachTo: App
                    }
                );
        });

        afterEach(() => {
            onSelect.mockClear();
            webixUISpy.mockClear();
            wrapper.unmount();
        });

        it('React OK', () => {
            expect(wrapper.exists()).toBe(true);
        });

        it('Webix OK', () => {

            expect(webix.ui).toHaveBeenCalledTimes(1); 
            //dropdown element
            const dropdown = webixFromSelector('.webix_control.webix_el_combo');
            // expect(dropdown.$view.innerHTML).toMatchSnapshot();

            //popup select options element
            const selectlist = webixFromSelector('.webix_popup');
            // expect(selectlist.$view.innerHTML).toMatchSnapshot();

        });

        it('Re-Renders after Prop Update', () => {

            wrapper.setProps({options : ['a','b','c'], value : 'c'}, ()=>{

                expect(wrapper.exists()).toBe(true);
                expect(webix.ui).toHaveBeenCalledTimes(2); 

                const dropdown = webixFromSelector('.webix_control.webix_el_combo');
                // expect(dropdown.$view.innerHTML).toMatchSnapshot();

                const selectlist = webixFromSelector('.webix_popup');
                // expect(selectlist.$view.innerHTML).toMatchSnapshot();

            });

        });

        it('Triggers Callback on Webix Value Change', () => { //TEST PASSES But Snapshot isnt as Expected

            const dropdown = webixFromSelector('.webix_control.webix_el_combo');
            
            dropdown.setValue('item2');
            expect(onSelect).toHaveBeenCalledTimes(1);
            expect(dropdown.getValue()).toBe('item2');
            // expect(dropdown.$view.innerHTML).toMatchSnapshot();  //still shows value="item3" ... the initial value

        });


        it('Triggers Callback on Click', async () => { //TEST FAILS, Click Events Dont seem to Register

            const dropdown = webixFromSelector('.webix_control.webix_el_combo');
  
            document.querySelector('.webix_el_combo input').click();
            document.querySelectorAll('.webix_popup .webix_list_item')[1].click();
      
           await waitForExpect(() => {

              expect(onSelect).toHaveBeenCalledTimes(1);
              expect(dropdown.getValue()).toBe('item2');

           });

        });





    });


});
