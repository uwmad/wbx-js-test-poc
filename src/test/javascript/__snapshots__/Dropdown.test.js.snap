// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown Rendering Re-Renders after Prop Update 1`] = `"<div class=\\"webix_el_box\\" style=\\"width:100px; height:40px\\"><input role=\\"combobox\\" placeholder=\\"placeholder text\\" id=\\"1546322400159\\" type=\\"combo\\" value=\\"c\\" style=\\"width: 96px; text-align: left;\\" aria-autocomplete=\\"list\\" aria-expanded=\\"false\\"><span style=\\"height:27px;padding-top:7px;\\" class=\\"webix_input_icon wxi-angle-down\\"></span></div>"`;

exports[`Dropdown Rendering Re-Renders after Prop Update 2`] = `"<div class=\\"webix_win_content\\" style=\\"padding: 0px;\\"><div class=\\"webix_win_head\\" style=\\"display: none;\\"></div><div class=\\"webix_win_body\\"><div class=\\"webix_view webix_list\\" role=\\"listbox\\" style=\\"overflow-y: hidden; width: 300px; height: 108px;\\" view_id=\\"$suggest4_list\\"><div class=\\"webix_scroll_cont\\"><div webix_l_id=\\"a\\" class=\\"webix_list_item\\" style=\\"width:auto; height:36px; overflow:hidden;\\" role=\\"option\\" tabindex=\\"0\\">a</div><div webix_l_id=\\"b\\" class=\\"webix_list_item\\" style=\\"width:auto; height:36px; overflow:hidden;\\" role=\\"option\\" tabindex=\\"-1\\">b</div><div webix_l_id=\\"c\\" class=\\"webix_list_item\\" style=\\"width:auto; height:36px; overflow:hidden;\\" role=\\"option\\" tabindex=\\"-1\\">c</div></div></div></div></div>"`;

exports[`Dropdown Rendering Webix OK 1`] = `"<div class=\\"webix_el_box\\" style=\\"width:100px; height:40px\\"><input role=\\"combobox\\" placeholder=\\"placeholder text\\" id=\\"1546322400083\\" type=\\"combo\\" value=\\"item3\\" style=\\"width: 96px; text-align: left;\\" aria-autocomplete=\\"list\\" aria-expanded=\\"false\\"><span style=\\"height:27px;padding-top:7px;\\" class=\\"webix_input_icon wxi-angle-down\\"></span></div>"`;

exports[`Dropdown Rendering Webix OK 2`] = `"<div class=\\"webix_win_content\\" style=\\"padding: 0px;\\"><div class=\\"webix_win_head\\" style=\\"display: none;\\"></div><div class=\\"webix_win_body\\"><div class=\\"webix_view webix_list\\" role=\\"listbox\\" style=\\"overflow-y: hidden; width: 300px; height: 144px;\\" view_id=\\"$suggest2_list\\"><div class=\\"webix_scroll_cont\\"><div webix_l_id=\\"item1\\" class=\\"webix_list_item\\" style=\\"width:auto; height:36px; overflow:hidden;\\" role=\\"option\\" tabindex=\\"0\\">item1</div><div webix_l_id=\\"item2\\" class=\\"webix_list_item\\" style=\\"width:auto; height:36px; overflow:hidden;\\" role=\\"option\\" tabindex=\\"-1\\">item2</div><div webix_l_id=\\"item3\\" class=\\"webix_list_item\\" style=\\"width:auto; height:36px; overflow:hidden;\\" role=\\"option\\" tabindex=\\"-1\\">item3</div><div webix_l_id=\\"item4\\" class=\\"webix_list_item\\" style=\\"width:auto; height:36px; overflow:hidden;\\" role=\\"option\\" tabindex=\\"-1\\">item4</div></div></div></div></div>"`;
